stackable-certs/src/ca/mod.rs

```diff
@@ -0,0 +1,359 @@
+    #[instrument(name = "create_certificate_authority_with", skip(signing_key_pair))]
+    pub fn new_with(signing_key_pair: S, serial_number: u64, validity: Duration) -> Result<Self> {
+        let serial_number = SerialNumber::from(serial_number);
+        let validity = Validity::from_now(*validity).context(ParseValiditySnafu)?;
+
+        // We don't allow customization of the CA subject by callers. Every CA
+        // created by us should contain the same subject consisting a common set
+        // of distinguished names (DNs).
+        let subject = Name::from_str(ROOT_CA_SUBJECT).context(ParseSubjectSnafu {
+            subject: ROOT_CA_SUBJECT,
+        })?;
+
+        let spki_pem = signing_key_pair
+            .verifying_key()
+            .to_public_key_pem(LineEnding::default())
+            .context(SerializePublicKeySnafu)?;
+
+        let spki = SubjectPublicKeyInfoOwned::from_pem(spki_pem.as_bytes())
+            .context(DecodeSpkiFromPemSnafu)?;
+
+        // There are multiple default extensions included in the profile. For
+        // the root profile, these are:
+        //
+        // - BasicConstraints marked as critical and CA = true
+        // - SubjectKeyIdentifier with the 160-bit SHA-1 hash of the subject
+        //   public key.
+        // - KeyUsage with KeyCertSign and CRLSign bits set. Ideally we also
+        //   want to include the DigitalSignature bit, which for example is
+        //   required for CA certs which want to sign an OCSP response.
+        //   Currently, the root profile doesn't include that bit.
+        //
+        // The root profile doesn't add the AuthorityKeyIdentifier extension.
+        // We manually add it below by using the 160-bit SHA-1 hash of the
+        // subject pulic key. This conforms to one of the outlined methods for
+        // generating key identifiers outlined in RFC 5280, section 4.2.1.2.
+        //
+        // Prepare extensions so we can avoid clones.
+        let aki = AuthorityKeyIdentifier::try_from(spki.owned_to_ref())
+            .context(ParseAuthorityKeyIdentifierSnafu)?;
+
+        let signer = signing_key_pair.signing_key();
+        let mut builder = CertificateBuilder::new(
+            Profile::Root,
+            serial_number,
+            validity,
+            subject,
+            spki,
+            signer,
+        )
+        .context(CreateCertificateBuilderSnafu)?;
+
+        // Add extension constructed above
+        builder
+            .add_extension(&aki)
+            .context(AddCertificateExtensionSnafu)?;
+
+        info!("create and sign CA certificate");
+        let certificate = builder.build().context(BuildCertificateSnafu)?;
+
+        Ok(Self {
+            serial_numbers: HashSet::new(),
+            certificate_pair: CertificatePair {
+                key_pair: signing_key_pair,
+                certificate,
+            },
+        })
+    }
+
+    #[instrument(skip(key_pair))]
+    pub fn generate_leaf_certificate<T>(
```

Comment:
**suggestion (non-blocker)**: `pub fn` should probably have doc-comments.

Reply:
Yes, I will add doc comments for it.